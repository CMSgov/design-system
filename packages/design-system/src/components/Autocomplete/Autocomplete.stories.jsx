import React from 'react';
import Autocomplete from './Autocomplete';
import TextField from '../TextField/TextField';
import { Title, Subtitle, Description, ArgsTable, PRIMARY_STORY } from '@storybook/addon-docs';

export default {
  title: 'Components/AutoComplete',
  component: Autocomplete,
  argTypes: {
    clearInputText: {
      control: { type: 'text' },
      type: { name: 'string', required: true },
    },
    // don't need to show empty textfield row
    textField: {
      table: { disable: true },
    },
    // items is an array, controls can be wonky
    items: { control: false },
  },
  args: {
    // setting some defaults so controls turn on by default
    clearInputText: 'Clear Search',
    clearInputOnBlur: true,
    clearSearchButton: true,
    loadingMessage: 'Loading...',
    noResultsMessage: 'No results',
  },
  subcomponents: { TextField },
  parameters: {
    docs: {
      page: () => (
        <>
          <Title />
          <Subtitle />
          <Description />
          <ArgsTable story={PRIMARY_STORY} />
        </>
      ),
    },
  },
};

const Template = (args) => {
  return (
    <Autocomplete
      {...args}
      // using downshifted props
      onChange={(evt, item) => console.log(item.selectedItem)}
      onInputValueChange={(evt, input) => console.log('[Autocomplete]: ' + input.inputValue)}
    >
      <TextField {...args.textField} name="Downshift_autocomplete" />
    </Autocomplete>
  );
};

export const LabeledList = Template.bind({});
LabeledList.args = {
  textField: {
    hint: 'Type c then use ARROW keys to change options, ENTER key to make a selection, ESC to dismiss.',
    label: 'Labeled list',
  },
  label: 'Select from the options below:',
  items: [
    {
      id: 'kRf6c2fY',
      name: 'Cook County, IL',
    },
    {
      id: 'lYf5cGfM',
      name: 'Cook County, MD',
    },
    {
      id: 'mZfKcGf9',
      name: 'Cook County, TN',
    },
    {
      id: 'xFz6dLba',
      name: 'Cook County, AK',
    },
    {
      id: 'vTr5c99',
      name: 'Cook County, FL',
    },
    {
      id: 'ntY8Lha',
      name: 'Cook County, AL',
    },
    {
      id: 'uRe0Wqo',
      name: 'Cook County, WA',
    },
    {
      id: 'yUR7MWl',
      name: 'Cook County, OR',
    },
  ],
};

export const CustomMarkup = Template.bind({});
CustomMarkup.args = {
  textField: {
    hint: 'Clicking the last item should not change the input value to "Search all snacks"',
    label: 'List with custom item markup',
  },
  label: 'Select from the options below:',
  items: [
    {
      id: '1',
      name: 'Carrots (1)',
      children: (
        <>
          Carrots <strong>(1)</strong>
        </>
      ),
    },
    {
      id: '2',
      name: 'Cookies (3)',
      children: (
        <>
          Cookies <strong>(3)</strong>
        </>
      ),
    },
    {
      id: '3',
      name: 'Crackers (2)',
      children: (
        <>
          Crackers <strong>(2)</strong>
        </>
      ),
    },
    {
      children: (
        <a href="#snacks" onClick={() => console.log('Searching for all the snacks!')}>
          Search all snacks
        </a>
      ),
      className: 'ds-c-autocomplete__search-all',
      isResult: false,
    },
  ],
};

export const LoadingMessage = Template.bind({});
LoadingMessage.args = {
  clearSearchButton: false,
  loading: true,
  items: [],
  textField: {
    hint: 'List should return string Loading to simulate async data call.',
    label: 'Loading Message',
  },
};

export const NoResults = Template.bind({});
NoResults.args = {
  items: [],
  clearSearchButton: false,
  textField: {
    hint: '',
    label: 'No Results Message',
  },
};
