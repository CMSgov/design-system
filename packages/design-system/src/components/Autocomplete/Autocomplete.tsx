import React, { useEffect, useRef } from 'react';
import Button from '../Button/Button';
import DropdownMenu from '../Dropdown/DropdownMenu';
import classNames from 'classnames';
import mergeRefs from '../utilities/mergeRefs';
import useId from '../utilities/useId';
import { errorPlacementDefault } from '../flags';
import {
  renderReactStatelyItems,
  renderStatusMessage,
  getTextFieldChild,
  getActiveDescendant,
  removeUndefined,
} from './utils';
import { t } from '../i18n';
import { useComboBox } from '../react-aria'; // from react-aria
import { useComboBoxState } from '../react-aria'; // from react-stately
import usePrevious from '../utilities/usePrevious';

export interface AutocompleteItem extends Omit<React.HTMLAttributes<'option'>, 'name'> {
  /**
   * Unique identifier for this item
   */
  id: string;
  /**
   * Displayed value of the item. May alternatively provide a `children` value
   */
  name?: string;
  /**
   * Custom React node as an alternative to a string-only `name`
   */
  children?: React.ReactNode;
  /**
   * Whether this item should be counted as one of the results for the purpose of announcing the
   * result count to screen readers
   * @deprecated This is no longer used, as we no longer have custom messaging for screen readers
   * @hide-prop [Deprecated]
   */
  isResult?: boolean;
}

export interface AutocompleteProps {
  /**
   * Screen reader-specific label for the Clear search `<button>`. Intended to provide a longer, more descriptive explanation of the button's behavior.
   */
  ariaClearLabel?: string;
  /**
   * Control the `TextField` autocomplete attribute. Defaults to "off" to support accessibility. [Read more.](https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion)
   */
  autoCompleteLabel?: string;
  /**
   * Must contain a `TextField` component
   */
  children: React.ReactNode;
  /**
   * Additional classes to be added to the root element.
   * Useful for adding utility classes.
   */
  className?: string;
  /**
   * Text rendered on the page if `clearInput` prop is passed. Default is "Clear search".
   */
  clearInputText?: React.ReactNode;
  /**
   * Removes the Clear search button when set to `false`
   */
  clearSearchButton?: boolean;
  /*
   * Sets the focus on the select during the first mount
   */
  autoFocus?: boolean;
  /**
   * @deprecated This is deprecated in favor of autoFocus
   * @hide-prop [Deprecated]
   */
  focusTrigger?: boolean;
  /**
   * A unique id to be passed to the child `TextField`. If no id is passed as a prop,
   * the `Autocomplete` component will auto-generate one. This prop was provided in cases
   * where an id might need to be passed to multiple components, such as the `htmlFor`
   * attribute on a label and the id of an input.
   */
  id?: string;
  /**
   * Customize the default status messages announced to screen reader users via aria-live when autocomplete results are populated.
   * @deprecated This is no longer used
   * @hide-prop [Deprecated]
   */
  getA11yStatusMessage?: any;
  /**
   * Access a reference to the child `TextField`'s `input` element
   */
  inputRef?: (...args: any[]) => any;
  /**
   * Used to determine the string value for the selected item (which is used to compute the `inputValue`).
   * @deprecated Please provide a `name` property to each item instead.
   * @hide-prop [Deprecated]
   */
  itemToString?: (item: AutocompleteItem) => string;
  /**
   * Array of objects used to populate the suggestion list that appears below the input as users type.
   * Passing an empty array will show a "No results" message. If you do not yet want to show results,
   * this props should be undefined.
   */
  items?: AutocompleteItem[];
  /**
   * Adds a heading to the top of the autocomplete list. This can be used to convey to the user that they're required to select an option from the autocomplete list.
   */
  label?: React.ReactNode;
  /**
   * A unique `id` to be used on the child `TextField` label tag
   */
  labelId?: string;
  /**
   * Can be called when the `items` array is being fetched remotely, or will be delayed for more than 1-2 seconds.
   */
  loading?: boolean;
  /**
   * Message users will see when the `loading` prop is passed to `Autocomplete`.
   */
  loadingMessage?: React.ReactNode;
  /**
   * Message users will see when the `items` array returns empty and the `loading` prop is passed to `<Autocomplete />`.
   */
  noResultsMessage?: React.ReactNode;
  /**
   * Called when the user selects an item and the selected item has changed. Called with the item that was selected.
   */
  onChange?: (selectedItem: AutocompleteItem) => void;
  /**
   * Called when the child `TextField` value changes. Is called with a string representing the input value.
   */
  onInputValueChange?: (inputValue: string) => void;
}

/**
 * The Autocomplete component wraps a TextField component and turns it into a combobox,
 * where a user can type into the text field and see matching results. They can then
 * select one of these results from the list, which will trigger an `onChange` event on
 * the Autocomplete.
 *
 * The two event handlers that should be used when this is a controlled component are
 * `onChange` and `onInputValueChange`. They are defined on the Autocomplete component
 * and not its child TextField component.
 *
 * As the user types and `onInputValueChange` is called, you should be supplying relevant
 * results to the Autocomplete through the `items` prop. The `items` prop is an array of
 * objects. Passing an empty array will show a "No results" message. If you do not yet
 * want to show results—for instance, because they haven't typed enough characters yet to
 * make a database call—the `items` prop should remain be undefined. If you are still
 * loading the results, use the `loading` boolean prop to display the loading message to
 * the user.
 *
 * For information about how and when to use this component,
 * [refer to its full documentation page](https://design.cms.gov/components/autocomplete/).
 */
export const Autocomplete = (props: AutocompleteProps) => {
  const id = useId('autocomplete--', props.id);
  const menuId = `${id}__menu`;

  const {
    ariaClearLabel,
    autoCompleteLabel,
    autoFocus,
    children,
    className,
    clearInputText,
    clearSearchButton,
    focusTrigger,
    inputRef: userInputRef,
    items,
    itemToString,
    label: menuHeading,
    labelId: menuHeadingId,
    loading,
    loadingMessage,
    noResultsMessage,
    onChange,
    onInputValueChange,
    ...autocompleteProps
  } = props;

  // Determine what we'll show based on state
  let reactStatelyItems = [];
  let statusMessage;
  if (items?.length) {
    reactStatelyItems = renderReactStatelyItems(items, itemToString);
  } else if (loading) {
    // If we're waiting for results to load, show the non-selected message
    statusMessage = renderStatusMessage(loadingMessage ?? t('autocomplete.loadingMessage'));
  } else if (items) {
    // If we have no results (empty array), show the non-selected message
    statusMessage = renderStatusMessage(noResultsMessage ?? t('autocomplete.noResultsMessage'));
  }

  const textField = getTextFieldChild(children);
  const size = textField.props.size;
  const labelId = textField.props.labelId ?? `${id}__label`;

  const state = useComboBoxState({
    ...autocompleteProps,
    allowsCustomValue: true,
    allowsEmptyCollection: true,
    children: reactStatelyItems,
    inputValue: textField.props.value,
    onInputChange: onInputValueChange
      ? (value) => {
          onInputValueChange(value);
        }
      : undefined,
    onSelectionChange: onChange
      ? (selectedKey) => {
          const selectedItem = items ? items.find((item) => selectedKey === item.id) : undefined;
          // We don't call onChange when the user deletes text, even though react-aria will call
          // this function with `null` if the input is cleared out. This is to maintain backwards
          // compatability, but we could consider changing this behavior in the future. If we
          // decide to remove this check, we can also remove the explicit `onChange` call in the
          // clear button handler.
          if (selectedItem) {
            onChange(selectedItem);
          }
        }
      : undefined,
  });

  const inputRef = useRef<HTMLInputElement>();
  const listBoxRef = useRef<HTMLElement>();
  const wrapperRef = useRef<HTMLDivElement>();
  const useComboboxProps = useComboBox(
    {
      ...autocompleteProps,
      name: textField.props.name,
      label: textField.props.label,
      inputRef,
      listBoxRef,
      popoverRef: listBoxRef,
    },
    state
  );

  // The display of bottom placed errorMessages in TextField breaks the Autocomplete's UI design.
  // Add errorMessageClassName to fix the styles for bottom placed errors
  const bottomError =
    (textField.props.errorPlacement === 'bottom' || errorPlacementDefault() === 'bottom') &&
    textField.props.errorMessage != null;

  const errorMessageClassName = classNames(
    textField.props.errorMessageClassName,
    bottomError && 'ds-c-autocomplete__error-message',
    bottomError && clearSearchButton && 'ds-c-autocomplete__error-message-clear-btn'
  );

  const textFieldProps = removeUndefined({
    ...useComboboxProps.inputProps,
    autoComplete: autoCompleteLabel,
    autoFocus: autoFocus || focusTrigger,
    'aria-activedescendant': useComboboxProps.inputProps['aria-activedescendant']
      ? getActiveDescendant(id, state, items)
      : undefined,
    'aria-controls': menuId,
    'aria-labelledby': labelId,
    errorMessageClassName,
    id,
    labelId,
    inputRef: mergeRefs([inputRef, userInputRef]),
    // Restores previous functionality where if you had typed characters into the text
    // field to get results and then blur away and come back, it'll open the results
    // list again without having to press anything on the keyboard.
    onFocus: (event) => {
      useComboboxProps.inputProps.onFocus?.(event);
      textField.props.onFocus?.(event);
      state.open();
    },
    // Allow the user to continue to attach their own event handlers to the TextField.
    // The following event handlers would normally be overwritten by useCombobox.
    onChange: (event) => {
      useComboboxProps.inputProps.onChange?.(event);
      textField.props.onChange?.(event);
    },
    onBlur: (event) => {
      useComboboxProps.inputProps.onBlur?.(event);
      textField.props.onBlur?.(event);
    },
    onTouchEnd: (event) => {
      useComboboxProps.inputProps.onTouchEnd?.(event);
      textField.props.onTouchEnd?.(event);
    },
    onKeyDown: (event) => {
      useComboboxProps.inputProps.onKeyDown?.(event);
      textField.props.onKeyDown?.(event);
    },
  });

  const rootClassName = classNames('ds-c-autocomplete', className);

  return (
    <div className={rootClassName} ref={wrapperRef}>
      {React.cloneElement(textField, textFieldProps)}

      {((state.isOpen && reactStatelyItems.length > 0) || (state.isFocused && statusMessage)) && (
        <DropdownMenu
          {...useComboboxProps.listBoxProps}
          componentClass="ds-c-autocomplete"
          heading={menuHeading}
          labelId={menuHeadingId}
          menuId={menuId}
          rootId={id}
          size={size}
          state={state}
          triggerRef={wrapperRef}
          listBoxRef={listBoxRef}
        >
          {statusMessage}
        </DropdownMenu>
      )}

      {clearSearchButton && (
        <Button
          aria-label={ariaClearLabel ?? t('autocomplete.ariaClearLabel')}
          className="ds-u-padding-right--0 ds-c-autocomplete__clear-btn"
          onClick={() => {
            state.setSelectedKey(null);
            state.setInputValue('');
            if (state.selectedKey) {
              onChange?.(null);
            }
          }}
          size="small"
          variation="ghost"
        >
          {clearInputText ?? t('autocomplete.clearInputText')}
        </Button>
      )}
    </div>
  );
};

Autocomplete.defaultProps = {
  autoCompleteLabel: 'off',
  clearSearchButton: true,
};

export default Autocomplete;
