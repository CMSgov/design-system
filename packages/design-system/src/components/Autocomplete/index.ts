export { default as Autocomplete } from './Autocomplete';
