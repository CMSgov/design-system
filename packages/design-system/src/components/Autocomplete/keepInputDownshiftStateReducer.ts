import { UseComboboxState, UseComboboxStateChangeOptions, useCombobox } from 'downshift';

/**
 * For some use cases of the Autocomplete component, we do not want the input to
 * be cleared when the input element loses focus. The default behavior of Downshift
 * is to clear the input value when the field loses focus. This custom Downshift
 * state reducer keeps that input value rather than clearing it.
 */
export default function keepInputDownshiftStateReducer(
  state: UseComboboxState<any>,
  actionAndChanges: UseComboboxStateChangeOptions<any>
) {
  const { changes, type } = actionAndChanges;
  switch (type) {
    case useCombobox.stateChangeTypes.InputBlur:
      return {
        ...changes,
        inputValue: state.inputValue,
        isOpen: false,
      };
    default:
      return changes;
  }
}
