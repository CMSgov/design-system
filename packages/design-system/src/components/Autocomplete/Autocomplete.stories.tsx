import React, { useState } from 'react';
import Autocomplete from './Autocomplete';
import TextField from '../TextField/TextField';
import uniqueId from 'lodash/uniqueId';
import { action } from '@storybook/addon-actions';
import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof Autocomplete> = {
  title: 'Components/Autocomplete',
  component: Autocomplete,
  args: {
    // setting some defaults so controls turn on by default
    clearInputText: 'Clear search',
    clearSearchButton: true,
    loadingMessage: 'Loading...',
    noResultsMessage: 'No results',
  } as any,
};
export default meta;

type Story = StoryObj<typeof Autocomplete>;

const Template = (args) => {
  const { items, textFieldLabel, textFieldHint, ...autocompleteArgs } = args;
  const [input, setInput] = useState('');
  const onInputValueChange = (...args) => {
    action('onInputValueChange')(args);
    setInput(args[0]);
  };
  let filteredItems = null;
  if (input.length > 0) {
    filteredItems = items.filter(
      (item) => !item.name || item.name.toLowerCase().includes(input.toLowerCase())
    );
  }
  return (
    <Autocomplete
      {...autocompleteArgs}
      onChange={action('onChange')}
      onInputValueChange={onInputValueChange}
      items={filteredItems}
    >
      <TextField label={textFieldLabel} hint={textFieldHint} name="autocomplete" value={input} />
    </Autocomplete>
  );
};

function makeItem(name: string, children?: React.ReactNode) {
  return {
    id: uniqueId(),
    name,
    children,
  };
}

export const Default: Story = {
  render: Template,
  args: {
    textFieldLabel: 'Enter and select a drug to see its cost under each plan',
    textFieldHint:
      'Type a letter to see results, then use ARROW keys to change options, ENTER key to make a selection, ESC to dismiss.',
    items: [
      makeItem('Acetaminophen'),
      makeItem('Advil'),
      makeItem('Benadryl'),
      makeItem('Claritin'),
      makeItem('Detrol'),
      makeItem('Excedrin'),
      makeItem('Flonase'),
      makeItem('Gas-X'),
      makeItem('Hydrocortisone'),
      makeItem('Ibuprofen'),
      makeItem('Insulin glargine'),
      makeItem('Jublia'),
      makeItem('Ketoprofen'),
      makeItem('Lidocaine'),
      makeItem('Lunestra'),
      makeItem('Melatonin'),
      makeItem('Meloxicam'),
      makeItem('Neosporin'),
      makeItem('Oxycodone'),
      makeItem('Prednisone'),
      makeItem('Prilosec'),
      makeItem('Quinine'),
      makeItem('Robitussin'),
      makeItem('Sudafed'),
      makeItem('Tamiflu'),
      makeItem('Unisom'),
      makeItem('Vicodin'),
      makeItem('Warfarin'),
      makeItem('Xanax'),
      makeItem('Yonsa'),
      makeItem('Zyrtec'),
    ],
  } as any,
};

export const LabeledList: Story = {
  render: Template,
  args: {
    textFieldLabel: 'Search for and select your county',
    textFieldHint:
      'Type "C" then use ARROW keys to change options, ENTER key to make a selection, ESC to dismiss.',
    label: 'Select from the options below:',
    items: [
      makeItem('Cook County, IL'),
      makeItem('Cook County, MD'),
      makeItem('Cook County, TN'),
      makeItem('Cook County, AK'),
      makeItem('Cook County, FL'),
      makeItem('Cook County, AL'),
      makeItem('Cook County, WA'),
      makeItem('Cook County, OR'),
    ],
  } as any,
};

export const CustomMarkup: Story = {
  render: Template,
  args: {
    textFieldLabel: 'Select a snack that starts with "C"',
    textFieldHint:
      'Type "C" to start seeing a list of snacks. Clicking the last item should not change the input value to "Search all snacks"',
    label: 'Select from the options below:',
    items: [
      makeItem(
        'Carrots (1)',
        <>
          Carrots <strong>(1)</strong>
        </>
      ),
      makeItem(
        'Cookies (3)',
        <>
          Cookies <strong>(3)</strong>
        </>
      ),
      makeItem(
        'Crackers (2)',
        <>
          Crackers <strong>(2)</strong>
        </>
      ),
      {
        children: (
          <a href="#snacks" onClick={() => console.log('Searching for all the snacks!')}>
            Search all snacks
          </a>
        ),
        className: 'ds-c-autocomplete__search-all',
        isResult: false,
      },
    ],
  } as any,

  parameters: {
    docs: {
      source: {
        code: 'Disabled for this story. See https://github.com/storybookjs/storybook/issues/11554',
      },
    },
  },
};

export const LoadingMessage: Story = {
  render: Template,
  args: {
    clearSearchButton: false,
    loading: true,
    items: [],
    textFieldLabel: 'This will only show a loading message',
    textFieldHint: 'List should return string Loading to simulate async data call.',
  } as any,
};

export const NoResults: Story = {
  render: Template,
  args: {
    items: [],
    clearSearchButton: false,
    textFieldLabel: 'This will show a "no results" message',
    textFieldHint: "Start typing, but you'll only get a loading message.",
  } as any,
};
