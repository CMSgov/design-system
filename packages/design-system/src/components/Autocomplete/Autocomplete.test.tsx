import React from 'react';
import Autocomplete from './Autocomplete';
import TextField from '../TextField/TextField';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const defaultItems = [{ id: 'kRf6c2fY', name: 'Cook County, IL' }];

function makeAutocomplete(customProps = {}) {
  const props = {
    items: defaultItems,
    children: <TextField label="autocomplete" name="autocomplete_field" />,
    id: 'static-id',
    ...customProps,
  };
  return <Autocomplete {...props} />;
}

function renderAutocomplete(customProps = {}) {
  return render(makeAutocomplete(customProps));
}

function open() {
  const autocompleteField = screen.getByRole('combobox');
  userEvent.click(autocompleteField);
  userEvent.type(autocompleteField, 'c');
}

function expectMenuToBeOpen() {
  expect(screen.getByRole('listbox')).toBeInTheDocument();
}

function expectMenuToBeClosed() {
  expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

describe('Autocomplete', () => {
  it('renders a closed Autocomplete component', () => {
    const { container } = renderAutocomplete();

    const wrapperEl = container.querySelectorAll('.ds-c-autocomplete');
    expect(wrapperEl.length).toEqual(1);

    const input = screen.getByRole('combobox');
    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute('aria-autocomplete', 'list');
    expect(input).toHaveAttribute('aria-controls');
    expect(input).toHaveAttribute('aria-expanded', 'false');
    expect(input).toHaveAttribute('role', 'combobox');
    expect(input).toHaveAttribute('type', 'text');

    const labelId = input.getAttribute('aria-labelledby');
    const label = container.querySelector(`#${labelId}`);
    expect(label).toBeInTheDocument();

    const menuEl = container.querySelector('.ds-c-autocomplete__menu-container');
    expect(menuEl).not.toBeInTheDocument();

    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
    expect(button.classList).toContain('ds-c-autocomplete__clear-btn');
    expect(button).toHaveAttribute('type', 'button');
    expect(button).toHaveAccessibleName('Clear search to try again');
  });

  it('renders items', () => {
    renderAutocomplete();

    open();
    expectMenuToBeOpen();

    const items = screen.getByRole('option');
    expect(items).toBeInTheDocument();
    expect(items).toHaveTextContent('Cook County, IL');
  });

  it('renders items with children property', () => {
    const items = [
      {
        id: '1',
        name: 'Carrots (1)',
        children: (
          <>
            Carrots <strong>(1)</strong>
          </>
        ),
      },
      {
        id: '2',
        name: 'Cookies (3)',
        children: (
          <>
            Cookies <strong>(3)</strong>
          </>
        ),
      },
      {
        id: '3',
        name: 'Crackers (2)',
        children: (
          <>
            Crackers <strong>(2)</strong>
          </>
        ),
      },
      {
        id: '4',
        name: 'Search all snacks',
        children: <a href="https://duckduckgo.com/?q=snacks">Search all snacks</a>,
      },
    ];

    renderAutocomplete({ items });
    open();
    const ul = screen.getByRole('listbox');
    expect(ul).toMatchSnapshot();
  });

  it('generates ids when no id is provided', () => {
    renderAutocomplete({ id: undefined });
    open();
    const idRegex = /autocomplete--\d+/;
    expect(screen.getByRole('listbox').id).toMatch(idRegex);
    expect(screen.getByRole('combobox').id).toMatch(idRegex);
  });

  it('renders item with custom className', () => {
    const items = [
      { id: '1a', name: 'Normal item' },
      { id: '5b', name: 'Special item', className: 'custom-class' },
    ];
    renderAutocomplete({ items });

    open();
    expectMenuToBeOpen();

    const listItems = screen.queryAllByRole('option');
    expect(listItems[1]).toHaveClass('custom-class');
    expect(listItems).toMatchSnapshot();
  });

  it('renders Autocomplete component without items', () => {
    renderAutocomplete({ items: undefined });
    open();
    expectMenuToBeClosed();
  });

  it('renders Autocomplete component no results', () => {
    renderAutocomplete({ items: [] });
    open();
    expect(screen.queryByRole('listbox').children.length).toEqual(1);
    expect(screen.queryByRole('option')).toHaveTextContent('No results');
  });

  it('shows the menu when open', () => {
    renderAutocomplete();
    open();
    expectMenuToBeOpen();
  });

  it('opens the menu when focusing on an input that has text in it', () => {
    renderAutocomplete({
      children: <TextField label="autocomplete" name="autocomplete_field" value="abc" />,
    });
    const autocompleteField = screen.getByRole('combobox');
    autocompleteField.focus();
    expectMenuToBeOpen();
  });

  it('opens the menu if typing resulted in results that were delayed by async data fetching', async () => {
    const { rerender } = renderAutocomplete({ items: undefined });
    const autocompleteField = screen.getByRole('combobox');
    userEvent.click(autocompleteField);
    userEvent.type(autocompleteField, 'ac');
    await sleep(100);
    rerender(makeAutocomplete({ items: defaultItems }));
    expectMenuToBeOpen();
  });

  it('does not render a clear search button when clearSearchButton is set to false', () => {
    renderAutocomplete({ clearSearchButton: false });

    expect(screen.queryByRole('button')).not.toBeInTheDocument();
  });

  it('renders custom class names', () => {
    const { container } = renderAutocomplete({ className: 'additional-class' });
    const child = container.querySelector('.ds-c-autocomplete');

    expect(child).toHaveClass('additional-class');
  });

  describe('default props', () => {
    it('defaults ariaClearLabel', () => {
      renderAutocomplete();
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();

      expect(button).toHaveAttribute('aria-label', 'Clear search to try again');
    });

    it('defaults clearInputText', () => {
      renderAutocomplete();
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();

      expect(button).toHaveTextContent('Clear search');
    });
  });

  it('should set the input value correctly when a listbox selection is clicked', () => {
    const onChange = jest.fn();
    renderAutocomplete({ onChange });
    const autocompleteField = screen.getByRole('combobox') as HTMLInputElement;
    userEvent.click(autocompleteField);
    userEvent.type(autocompleteField, 'c');

    const listboxItem = screen.getByRole('option');
    userEvent.click(listboxItem);

    expect(autocompleteField.value).toBe('Cook County, IL');
    expect(onChange).toHaveBeenCalledWith(defaultItems[0]);
    expectMenuToBeClosed();
  });

  it('should set the input value to empty when "Clear search" is clicked', () => {
    renderAutocomplete();
    const autocompleteField = screen.getByRole('combobox') as HTMLInputElement;
    userEvent.click(autocompleteField);
    userEvent.type(autocompleteField, 'c');

    const listboxItem = screen.getByRole('option');
    userEvent.click(listboxItem);

    const clearButton = screen.getByText('Clear search');
    userEvent.click(clearButton);

    expect(autocompleteField.value).toBe('');
  });

  it('should call onChange with null item when "Clear search" is clicked', () => {
    const onChange = jest.fn();
    renderAutocomplete({ onChange });
    const autocompleteField = screen.getByRole('combobox') as HTMLInputElement;
    userEvent.click(autocompleteField);
    userEvent.type(autocompleteField, 'c');

    const listboxItem = screen.getByRole('option');
    userEvent.click(listboxItem);

    const clearButton = screen.getByText('Clear search');
    userEvent.click(clearButton);

    expect(autocompleteField.value).toBe('');
    expect(onChange).toHaveBeenLastCalledWith(null);
  });

  it('should select list items by keyboard', () => {
    const onChange = jest.fn();
    renderAutocomplete({ onChange });
    const autocompleteField = screen.getByRole('combobox') as HTMLInputElement;
    userEvent.click(autocompleteField);
    userEvent.type(autocompleteField, 'c');
    userEvent.type(autocompleteField, '{arrowdown}');
    userEvent.type(autocompleteField, '{enter}');

    expect(autocompleteField.value).toBe('Cook County, IL');
    expect(onChange).toHaveBeenCalledWith(defaultItems[0]);
    expectMenuToBeClosed();
  });

  it('should not call onChange when an item was not selected', () => {
    const onChange = jest.fn();
    renderAutocomplete({ onChange });
    const autocompleteField = screen.getByRole('combobox') as HTMLInputElement;
    userEvent.click(autocompleteField);
    userEvent.type(autocompleteField, 'c');
    userEvent.type(autocompleteField, 'o');
    userEvent.type(autocompleteField, '{enter}');
    expect(onChange).not.toHaveBeenCalled();
  });

  it('should clear the input value by keyboard', () => {
    renderAutocomplete();
    const autocompleteField = screen.getByRole('combobox') as HTMLInputElement;
    autocompleteField.focus();
    userEvent.click(autocompleteField);
    userEvent.type(autocompleteField, 'c');
    userEvent.type(autocompleteField, '{arrowdown}');
    userEvent.type(autocompleteField, '{enter}');

    expect(autocompleteField.value).toBe('Cook County, IL');

    userEvent.tab();

    const clearButton = screen.getByText('Clear search');
    userEvent.click(clearButton);

    expect(autocompleteField.value).toBe('');
  });

  it('closes the listbox when ESC is pressed', () => {
    renderAutocomplete();
    const autocompleteField = screen.getByRole('combobox') as HTMLInputElement;
    userEvent.click(autocompleteField);
    userEvent.type(autocompleteField, 'c');

    expectMenuToBeOpen();

    expect(autocompleteField.value).toEqual('c');

    userEvent.type(autocompleteField, '{esc}');

    expectMenuToBeClosed();
  });

  it("calls child TextField's event handlers", () => {
    const props = {
      label: 'autocomplete',
      name: 'autocomplete_field',
      onFocus: jest.fn(),
      onChange: jest.fn(),
      onKeyDown: jest.fn(),
      // onTouchEnd : jest.fn(), Doesn't look like we can actually test onTouchEnd
      onBlur: jest.fn(),
    };
    renderAutocomplete({ children: <TextField {...props} /> });
    const field = screen.getByRole('combobox');
    userEvent.click(field);
    expect(props.onFocus).toHaveBeenCalledTimes(1);
    userEvent.type(field, 'c');
    expect(props.onKeyDown).toHaveBeenCalledTimes(1);
    expect(props.onChange).toHaveBeenCalledTimes(1);
    userEvent.tab();
    expect(props.onBlur).toHaveBeenCalledTimes(1);
  });

  it('allows arbitrary props to remain on the TextField', () => {
    const props = {
      label: 'autocomplete',
      name: 'autocomplete_field',
      placeholder: 'Hello world!',
      fieldClassName: 'a-custom-class',
    };
    renderAutocomplete({ children: <TextField {...props} /> });
    const field = screen.getByRole('combobox');
    expect(field).toHaveAttribute('placeholder', props.placeholder);
    expect(field).toHaveClass(props.fieldClassName);
  });

  it('inherits size prop from the TextField', () => {
    const { container } = renderAutocomplete({
      children: <TextField label="autocomplete" name="field" size="medium" />,
    });
    const field = screen.getByRole('combobox');
    expect(field).toHaveClass('ds-c-field--medium');
    open();
    const menuContainer = container.querySelector('.ds-c-autocomplete__menu-container');
    expect(menuContainer).toHaveClass('ds-c-field--medium');
  });
});
