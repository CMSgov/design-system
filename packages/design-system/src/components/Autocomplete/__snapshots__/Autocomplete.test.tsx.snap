// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete renders item with custom className 1`] = `
Array [
  <li
    aria-selected="false"
    class="ds-c-autocomplete__menu-item"
    data-key="1a"
    id="static-id__item--0"
    role="option"
  >
    Normal item
  </li>,
  <li
    aria-selected="false"
    class="custom-class ds-c-autocomplete__menu-item"
    data-key="5b"
    id="static-id__item--1"
    role="option"
  >
    Special item
  </li>,
]
`;

exports[`Autocomplete renders items with children property 1`] = `
<ul
  aria-label="Suggestions"
  aria-labelledby=""
  class="ds-c-autocomplete__menu"
  id="static-id__menu"
  role="listbox"
>
  <li
    aria-selected="false"
    class="ds-c-autocomplete__menu-item"
    data-key="1"
    id="static-id__item--0"
    role="option"
  >
    Carrots 
    <strong>
      (1)
    </strong>
  </li>
  <li
    aria-selected="false"
    class="ds-c-autocomplete__menu-item"
    data-key="2"
    id="static-id__item--1"
    role="option"
  >
    Cookies 
    <strong>
      (3)
    </strong>
  </li>
  <li
    aria-selected="false"
    class="ds-c-autocomplete__menu-item"
    data-key="3"
    id="static-id__item--2"
    role="option"
  >
    Crackers 
    <strong>
      (2)
    </strong>
  </li>
  <li
    aria-selected="false"
    class="ds-c-autocomplete__menu-item"
    data-key="4"
    id="static-id__item--3"
    role="option"
  >
    <a
      href="https://duckduckgo.com/?q=snacks"
    >
      Search all snacks
    </a>
  </li>
</ul>
`;
